// @redocly/openapi-core/lib/rules/oas3/spec-components-invalid-map-name.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpecComponentsInvalidMapName = void 0;
const SpecComponentsInvalidMapName = () => {
    const KEYS_REGEX = '^[a-zA-Z0-9\\.\\-_]+$';
    function validateKey(key, report, location, component) {
        if (!new RegExp(KEYS_REGEX).test(key)) {
            report({
                message: `The map key in ${component} "${key}" does not match the regular expression "${KEYS_REGEX}"`,
                location: location.key(),
            });
        }
    }
    return {
        Components: {
            Parameter(_node, { key, report, location }) {
                validateKey(key, report, location, 'parameters');
            },
            Response(_node, { key, report, location }) {
                validateKey(key, report, location, 'responses');
            },
            Schema(_node, { key, report, location }) {
                validateKey(key, report, location, 'schemas');
            },
            Example(_node, { key, report, location }) {
                validateKey(key, report, location, 'examples');
            },
            RequestBody(_node, { key, report, location }) {
                validateKey(key, report, location, 'requestBodies');
            },
            Header(_node, { key, report, location }) {
                validateKey(key, report, location, 'headers');
            },
            SecurityScheme(_node, { key, report, location }) {
                validateKey(key, report, location, 'securitySchemas');
            },
            Link(_node, { key, report, location }) {
                validateKey(key, report, location, 'links');
            },
            Callback(_node, { key, report, location }) {
                validateKey(key, report, location, 'callbacks');
            },
        },
    };
};
exports.SpecComponentsInvalidMapName = SpecComponentsInvalidMapName;
