// docusaurus-theme-openapi-docs/lib-next/theme/ApiItem/index.js
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
import React from "react";
import BrowserOnly from "@docusaurus/BrowserOnly";
import ExecutionEnvironment from "@docusaurus/ExecutionEnvironment";
import { HtmlClassNameProvider } from "@docusaurus/theme-common";
import useDocusaurusContext from "@docusaurus/useDocusaurusContext";
import useIsBrowser from "@docusaurus/useIsBrowser";
import { Provider } from "react-redux";
import { createAuth } from "../ApiDemoPanel/Authorization/slice";
import { createPersistanceMiddleware } from "../ApiDemoPanel/persistanceMiddleware";
import DocItemLayout from "./Layout";
import DocItemMetadata from "./Metadata";
import { createStoreWithoutState, createStoreWithState } from "./store";

const { DocProvider } = require("@docusaurus/theme-common/internal");

let ApiDemoPanel = (_) => <div />;

if (ExecutionEnvironment.canUseDOM) {
  ApiDemoPanel = require("@theme/ApiDemoPanel").default;
}

export default function ApiItem(props) {
  const docHtmlClassName = `docs-doc-id-${props.content.metadata.unversionedId}`;
  const MDXComponent = props.content;
  const { frontMatter } = MDXComponent;
  const { info_path: infoPath } = frontMatter;
  const { api } = frontMatter;
  const { siteConfig } = useDocusaurusContext();
  const themeConfig = siteConfig.themeConfig;
  const options = themeConfig.api;
  const isBrowser = useIsBrowser(); // Define store2

  let store2 = {};
  const persistanceMiddleware = createPersistanceMiddleware(options); // Init store for SSR

  if (!isBrowser) {
    store2 = createStoreWithoutState({}, [persistanceMiddleware]);
  } // Init store for CSR to hydrate components

  if (isBrowser) {
    const acceptArray = Array.from(
      new Set(
        Object.values(api?.responses ?? {})
          .map((response) => Object.keys(response.content ?? {}))
          .flat()
      )
    );
    const content = api?.requestBody?.content ?? {};
    const contentTypeArray = Object.keys(content);
    const servers = api?.servers ?? [];
    const params = {
      path: [],
      query: [],
      header: [],
      cookie: [],
    };
    api?.parameters?.forEach((param) => {
      const paramType = param.in;
      const paramsArray = params[paramType];
      paramsArray.push(param);
    });
    const auth = createAuth({
      security: api?.security,
      securitySchemes: api?.securitySchemes,
      options,
    }); // TODO: determine way to rehydrate without flashing
    // const acceptValue = window?.sessionStorage.getItem("accept");
    // const contentTypeValue = window?.sessionStorage.getItem("contentType");

    const server = window?.sessionStorage.getItem("server");
    const serverObject = JSON.parse(server) ?? {};
    store2 = createStoreWithState(
      {
        accept: {
          value: acceptArray[0],
          options: acceptArray,
        },
        contentType: {
          value: contentTypeArray[0],
          options: contentTypeArray,
        },
        server: {
          value: serverObject.url ? serverObject : undefined,
          options: servers,
        },
        response: {
          value: undefined,
        },
        body: {
          type: "empty",
        },
        params,
        auth,
      },
      [persistanceMiddleware]
    );
  }

  if (api) {
    return (
      <DocProvider content={props.content}>
        <HtmlClassNameProvider className={docHtmlClassName}>
          <DocItemMetadata />
          <DocItemLayout>
            <Provider store={store2}>
              <div className="row">
                <div className="col col--7">
                  <MDXComponent />
                </div>
                <div className="col col--5">
                  <BrowserOnly fallback={<div>Loading...</div>}>
                    {() => {
                      return <ApiDemoPanel item={api} infoPath={infoPath} />;
                    }}
                  </BrowserOnly>
                </div>
              </div>
            </Provider>
          </DocItemLayout>
        </HtmlClassNameProvider>
      </DocProvider>
    );
  } // Non-API docs

  return (
    <DocProvider content={props.content}>
      <HtmlClassNameProvider className={docHtmlClassName}>
        <DocItemMetadata />
        <DocItemLayout>
          <div className="row">
            <div className="col col--12">
              <MDXComponent />
            </div>
          </div>
        </DocItemLayout>
      </HtmlClassNameProvider>
    </DocProvider>
  );
}
